Sources/ClientRuntime/Networking/Http/URLSession/URLSessionHTTPClient.swift

```diff
@@ -0,0 +1,313 @@
+        private let lock = NSRecursiveLock()
+
+        /// A dictionary of `Connection`s, keyed by the `URLSessionTask` associated with them.
+        private var connections = [URLSessionTask: Connection]()
+
+        /// Adds a connection to the storage, keyed by its `URLSessionTask`.
+        func set(_ connection: Connection, for key: URLSessionTask) {
+            lock.lock()
+            defer { lock.unlock() }
+            connections[key] = connection
+        }
+
+        /// Allows modification of a `Connection` while holding exclusive access to it.
+        ///
+        /// Do not keep a reference to the connection outside the scope of `block`, or modify the connection after `block` returns.
+        func modify(_ key: URLSessionTask, block: (Connection) -> Void) {
+            lock.lock()
+            defer { lock.unlock() }
+            guard let connection = connections[key] else { return }
+            block(connection)
+        }
+
+        /// Removes the connection keyed by `key` from storage.
+        func remove(_ key: URLSessionTask) {
+            lock.lock()
+            defer { lock.unlock() }
+            connections.removeValue(forKey: key)
+        }
+    }
+
+    /// Handles URLSession delegate callbacks.
+    private final class SessionDelegate: NSObject, URLSessionDataDelegate {
+
+        /// Holds connection records for all in-progress connections.
+        let storage = Storage()
+
+        /// Logger for HTTP-related events.
+        let logger: LogAgent
+
+        init(logger: LogAgent) {
+            self.logger = logger
+        }
+
+        /// Called when the initial response to a HTTP request is received.
+        /// This callback is made as soon as the initial response + headers is complete.
+        /// Response body data may continue to stream in after this callback is received.
+        func urlSession(
+            _ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse
+        ) async -> URLSession.ResponseDisposition {
+            logger.debug("urlSession(_:dataTask:didReceive:) called")
+            storage.modify(dataTask) { connection in
+                guard let httpResponse = response as? HTTPURLResponse else {
+                    logger.error("Received non-HTTP urlResponse")
+                    let error = URLSessionHTTPClientError.responseNotHTTP
+                    connection.continuation?.resume(throwing: error)
+                    connection.continuation = nil
+                    return
+                }
+                let statusCode = HttpStatusCode(rawValue: httpResponse.statusCode) ?? .insufficientStorage
+                let httpHeaders: [HTTPHeader] = httpResponse.allHeaderFields.compactMap { (name, value) in
+                    guard let name = name as? String else { return nil }
+                    return HTTPHeader(name: name, value: String(describing: value))
+                }
```

Comment:
I am not sure what the encoding of the value will be here. Is it possible that this could run into the same issue when trying to decode it? For example, if the encoding was `ISO-8859-1` but we are trying to decode it as `UTF-8`, which will fail.

Reply:
Foundation should convert the headers from `ISO-8859-1` as needed to form a Swift string.  So that step should have been handled before these header values are returned to us.

If you can show me an example where header encoding has caused bugs in the past, I can test it with this client.